lib/register-authenticator.js

```diff
@@ -0,0 +1,271 @@
+async function httpAuthenticate(logger, data, url, hook_method, secret, username, password, req) {
+  const {AlertType, writeAlerts} = req.srf.locals;
+  const {account_sid} = req.locals;
+  try {
+    let uri = url;
+    let body;
+    const method = hook_method ? hook_method.toUpperCase() : 'POST';
+    if ('GET' === method) {
+      const str = qs.stringify(data);
+      uri = `${uri}?${str}`;
+    } else {
+      body = data;
+    }
+    const headers = {
+      ...(username &&
+        password &&
+        basicAuth(username, password)),
+      ...(secret && generateSigHeader(body || 'null', secret))
+    };
+    const request = bent(
+      'json',
+      200,
+      method,
+      headers
+    );
+    const json = await request(uri, body, headers);
+    return {
+      ...json,
+      statusCode: 200
+    };
+  } catch (err) {
+    logger.info(`Error from calling auth callback: ${err}`);
+    let opts = { account_sid };
+    if (err.code === 'ECONNREFUSED') {
+      opts = { ...opts, alert_type: AlertType.WEBHOOK_CONNECTION_FAILURE, url: err.hook };
+    }
+    else if (err.code === 'ENOTFOUND') {
+      opts = { ...opts, alert_type: AlertType.WEBHOOK_CONNECTION_FAILURE, url: err.hook };
+    }
+    else if (err.name === 'StatusError') {
+      opts = { ...opts, alert_type: AlertType.WEBHOOK_STATUS_FAILURE, url: err.hook, status: err.statusCode };
+    }
+    if (opts.alert_type) {
+      try {
+        await writeAlerts(opts);
+      } catch (err) {
+        logger.error({ err, opts }, 'Error writing alert');
+      }
+    }
+
+    return {
+      status: 'failed',
+      statusCode: err.statusCode || 500
+    };
+  }
+}
+
+function calculateResponse({username, realm, method, nonce, uri, nc, cnonce, qop}, password) {
+  const ha1 = crypto.createHash('md5');
+  ha1.update([username, realm, password].join(':'));
+  const ha2 = crypto.createHash('md5');
+  ha2.update([method, uri].join(':'));
+
+  // Generate response hash
+  const response = crypto.createHash('md5');
+  const responseParams = [
+    ha1.digest('hex'),
+    nonce
+  ];
+
+  if (cnonce) {
+    responseParams.push(nc);
+    responseParams.push(cnonce);
+  }
+
+  if (qop) {
+    responseParams.push(qop);
+  }
+  responseParams.push(ha2.digest('hex'));
+  response.update(responseParams.join(':'));
+
+  return response.digest('hex');
+}
+
+async function clientAuthentication(logger, data, req) {
+  const {username, response} = data;
+  const {account_sid} = req.locals;
+  const {lookupClientByAccountAndUsername} = req.srf.locals.dbHelpers;
+
+  const clients = await lookupClientByAccountAndUsername(account_sid, username);
+  if (clients.length) {
+    // Only take the first result.
+    const client = clients[0];
+    if (calculateResponse(data, client.password) === response) {
+      return {
+        status: 'ok',
+        statusCode: 200
+      };
+    }
+  }
+  return {
+    status: 'failed',
+    statusCode: 200
+  };
+}
+
+const digestChallenge = async(req, res, next) => {
```

Comment:
I'm wondering if we should refactor these functions into a new package.  Do we have this exact code elsewhere (ie duplicated?)